src/traits-canonicalization.md

```diff
@@ -0,0 +1,93 @@
+# Canonicalization
+
+Canonicalization is the process of **isolating** an inference value
+from its context. It is really based on a very simple concept: every
+[inference variable](./type-inference.html#vars) is always in one of two
+states: either it is **unbound**, in which case we don't know yet what
+type it is, or it is **bound**, in which case we do. So to isolate
+some thing T from its environment, we just walk down and find the
+unbound variables that appear in T; those variables get renumbered in
+a canonical order (left to right, for the most part, but really it
+doesn't matter as long as it is consistent).
+
+So, for example, if we have the type `X = (?T, ?U)`, where `?T` and
+`?U` are distinct, unbound inference variables, then the canonical
+form of `X` would be `(?0, ?1)`, where `?0` and `?1` represent these
+**canonical placeholders**. Note that the type `Y = (?U, ?T)` also
+canonicalizes to `(?0, ?1)`. But the type `Z = (?T, ?T)` would
+canonicalize to `(?0, ?0)` (as would `(?U, ?U)`). In other words, the
+exact identity of the inference variables is not important -- unless
+they are repeated.
+
+We use this to improve caching as well as to detect cycles and other
+things during trait resolution. Roughly speaking, the idea is that if
+two trait queries have the same canonicalize form, then they will get
+the same answer -- modulo the precise identities of the variables
+involved.
+
+To see how it works, imagine that we are asking to solve the following
+trait query: `?A: Foo<'static, ?B>`, where `?A` and `?B` are unbound.
+This query contains two unbound variables, but it also contains the
+lifetime `'static`. The trait system generally ignores all lifetimes
+and treats them equally, so when canonicalizing, we will *also*
+replace any [free lifetime](./background.html#free-vs-bound) with a
```

Comment:
Yes. Re: skolemization, there is a strong theoretical relationship -- but there is also a difference. With skolemization -- at least how we are using it elsewhere -- we are trying to prove that the query holds **for any lifetime** (or type, whatever). In this case, we are not requiring that it hold for any lifetime; instead, we propagate back requirements on the input lifetimes. So for example it might be that you have `impl Foo<T> for T`, and hence `&'a u32: Foo<&'b u32>` only if `'a == 'b`. We would propagate back the requirement that `'a == 'b` as part of the query result (but we would not check it).

Reply:
Ah, that makes sense :+1: 

Thanks!